Implemented Schema Display and removed tooltip, Added SVG Icon name before them

nomandhoni-cs opened the pull request:
## Commit 2ac8390bca5a63c649ca9f2dceaebe6759cf5774
#### Implemented Schema Display and removed old way of displaying by tooltip, 
- This will not increase the size of generated html because of displaying single div dynamically
#### Changes
- Used a separate div for display the schema
- When a clicked on the accordion, I insert the same schema display div in the accordion dynamically
- You can one display for all. That's why this will not increase the size of the generated html file
#### Demo Video
https://user-images.githubusercontent.com/92979541/231662325-687fbbcc-cc79-4d87-9027-fa467def8a11.mp4

## Commit c62dc94bfc917541531db85c0d23f770b11ea398
#### Added Icon name before SVG to gain more readability
#### Why I Did this ?
When going through these svg, I have to find out the class and based on the class I have guess this icon is look like this.
This is a kind of hassle. That's why added the Icon name before the icon by comment.

Julian commented:
Nice. Thanks! I'll have a closer look and leave some more specific comments, but at first glance the most obvious thing that stands out is that you're inserting *instances* into the box for schemas.

E.g. in your video, `{"type": "integer"}` is a schema -- it's the schema for all the examples you're clicking on.

37 (or "37" or etc.) are not schemas, they don't belong in the box labelled schema -- they're *instances*, data that the schema is validating. So someone wants to see both of those at the same time (the schema, and the instance that's valid or invalid under it)

We need to think about where the best place to render the instances is if we're going to do that dynamically -- maybe it makes sense to split your schema box into 2 columns, put the schema on the left, and dynamically change the instance on the right?

nomandhoni-cs commented on bowtie/templates/report.html.j2:
```diff
@@ -47,6 +50,10 @@
     font-size: 0.8rem;
   }
+  {# Schema diplay hidden default  #}
+  #display-body{
```
Sure! I missed that bootstrap class.
Thanks

Julian replied:
You resolved the comment but this looks like it's still in the diff, no?

nomandhoni-cs replied:
Opps! There was a mistake. First Time I made a new commit with every thing but I forgot to git pull before commiting new changes. Then I had to stash that commit and rewrite these. In that time I forgot to remove this css.

nomandhoni-cs replied:
Done! Used Bootstrap class ( Thanks for correcting me )

nomandhoni-cs commented:
That's great, thanks for correcting me. 
Let's see which one looks good, showing instances in a column or in a separate row below. With title 'Instance' above.

Julian commented:
> There are some schemas, too wide, I think they're gonna take whole space.

I think we'll find there are both schemas *and* instances which are going to be too wide, so no matter what we'll need to handle that.

But I'm ok seeing how it looks with a horizontal split which it sounds like is what you're suggesting to start -- so schema on top and last-selected instance on bottom? Probably it's nice to highlight the row of the instance that's shown if we do that.

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -2,6 +2,9 @@
 
 <style>
+  a:link{
+    text-decoration: none;
```
Here too -- can you elaborate on which link you're adding this for? I would also suspect Bootstrap should have some class for this though I didn't check.

nomandhoni-cs replied:
Currently There are no need to use `<a href="#">` for these links schema and test name, I just wanted to remove default link style (Underline)
![Screenshot_90](https://user-images.githubusercontent.com/92979541/231989777-d11ca1d6-b43e-4556-8ff6-21bfabdcb707.png)

nomandhoni-cs replied:
Done! Used bootstrap class instead of hard coded css

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -410,13 +419,93 @@
+        displayBody.classList.add("d-block");
+        displayCode(code);
+        const isDisplayed = displayBody.parentElement === document.getElementById(parentBodyId);
+        // If the code segment is displayed, remove it
```
A lot of the comments above are fine/definitely improvements for understanding SVG icons, but these kind of comments seem less so, they just repeat what some fairly simple code seems to be doing, or repeat the name of a function.

nomandhoni-cs replied:
Removed some of the comments

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -410,13 +419,93 @@
+                                                            </svg>`;
+setTimeout(() => {
+  {# Clipboard SVG Icon setting as a default again after copy  #}
+  document.getElementById("copy-button").innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16">
```
I think rather than continuing to copy paste SVG elements, let's just use [Bootstrap Icons](https://icons.getbootstrap.com/#icon-font) at this point? Which I think already has most of the SVGs both that were already present as well as the new ones you're adding.

nomandhoni-cs replied:
In the html report, every SVG icon is from bootstrap. Yes we can Directly use CDN for icons and, Instead of using the svg code we will use icons like `<i class="bi bi-0-circle"></i>` this.

Julian replied:
Ok let's decide in a follow-up and stick as-is for now.

nomandhoni-cs commented:
Hey @Julian,
I think this pull request is ready to be merged!
- Solved my silly mistakes ( Using hard coded css instead of using bootstrap class )
- Implemented display schema and instance info display in the same row
- Functional copy button
- Removed close button ( I Will add this later, currently thinking what is the best way to do this with less code )
![Screenshot_92](https://user-images.githubusercontent.com/92979541/232242083-dde4031e-7ab0-4876-9cbc-59f5a154c137.png)
 

https://user-images.githubusercontent.com/92979541/232242189-71d1a979-a212-40f0-854e-8baa4ef77c32.mp4

## Functionality I will bring in next PR
- Highlight the clicked instance
- Close button
- Find out more enhancement

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -410,13 +413,100 @@
+        document.body.appendChild(tempElement);
+        tempElement.select();
+        tempElement.setSelectionRange(0, 99999);
+        document.execCommand("copy");
```
I think this is the old API for doing this -- on not-super-old browsers I think you can just use [the clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API).

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -278,8 +282,7 @@
                 <td>
-                  <a href="#" data-bs-toggle="tooltip" data-bs-html="true"
-                    data-bs-title='<pre>{{ test.instance | tojson(indent=2) }}</pre>'>{{ test.description }}</a>
+                  <a href="#schema" class="text-decoration-none" onclick='displayCode(`{{ test.instance | tojson(indent=2) }}`,"instance-info")'>{{ test.description }}</a>
```
I think this `onClick` should go higher -- right now it only renders the instance if you click on the text, but someone is likely to click on the *row*, even beyond the text. So maybe on the `tr`? Until/unless we have some way of displaying the actual results more clearly.

Julian commented:
Thanks Noman! Left a few comments directly on the PR. Two others:

* Something's broken when collapsing accordion items -- see e.g.:

https://user-images.githubusercontent.com/329822/232501371-edab2165-2b46-40b5-9ab6-1c2b80006efe.mov

where after collapsing one, reopening it doesn't render the schema and instance anymore.

* Can you rebase or merge the main branch into yours, it bumps the bootstrap version so let's just be sure it still works after your changes

nomandhoni-cs commented:
Thanks, I didn't think of that testing, Closing, and opening the same accordion. I the I have to save the state with boolean and based on that I have to execute the function.

nomandhoni-cs commented:
Thanks a lot @Julian,
There was this little mistake, that I forgot to handle the edge case, that if I click on the same accordion for open and close.
- Now handled that with condition to see if the accordion already open or not
- Based on that execute the function
https://user-images.githubusercontent.com/92979541/232547118-013ec25b-e25a-45cd-a88a-e2baebf980ea.mp4

Julian commented:
Great -- I see you have a few more minor style issues -- I just added a linter to try and catch some, though I'm not sure it'll catch all of them yet. But can you again merge main into your branch and see if you can fix some of the remaining spacing / style issues?

nomandhoni-cs commented:
Merged from my main branch to fix spacing / style issue as you told me to do.

Julian commented on bowtie/templates/report.html.j2:
```diff
@@ -1,6 +1,54 @@
 <html lang="en">
 
+<style>
+  .details-required {
```
This might be a merge mistake -- the style tag belongs in the head tag if we're going to have it -- though a few more things below look like they should be doable with normal bootstrap classes -- but can you move all this into the style tag in the head tag?

nomandhoni-cs replied:
and Done!
I cannot express my gratitude enough for your incredible patience throughout this journey. Your unwavering support and guidance have been invaluable to me. Your help in identifying and correcting my mistakes has been particularly meaningful. Thank you so much for everything.

Julian replied:
No worries at all, definitely not a race! Thanks for the efforts!

Julian commented:
Looks nice! Merging, really appreciated! Will deploy in just a bit.
